pkg/client/bigquery.go

```diff
@@ -0,0 +1,202 @@
+		client:    client,
+	}, nil
+}
+
+// Close releases any resources held by the BigQuery client.
+func (bq *BigQuery) Close() {
+	bq.client.Close()
+}
+
+// Check if an entry with a given delivery_id already exists in the events table, this attempts to prevent duplicate processing of events.
+// This is used by the webhook service.
+func (bq *BigQuery) DeliveryEventExists(ctx context.Context, eventsTableID, deliveryID string) (bool, error) {
+	res, err := bq.makeCountQuery(ctx, fmt.Sprintf("SELECT COUNT(1) FROM `%s.%s.%s` WHERE delivery_id = '%q'",
+		bq.projectID,
+		bq.datasetID,
+		eventsTableID,
+		deliveryID,
+	))
+	if err != nil {
+		return false, fmt.Errorf("failed to execute DeliveryEventExists: %w", err)
+	}
+
+	return res > 0, nil
+}
+
+// Check if the number of entries with a given delivery_id in the failure-events table exceeds the retry limit.
+// This is used by the webhook service.
+func (bq *BigQuery) FailureEventsExceedsRetryLimit(ctx context.Context, failureEventTableID, deliveryID string, maxRetry int) (bool, error) {
+	res, err := bq.makeCountQuery(ctx, fmt.Sprintf("SELECT COUNT(1) FROM `%s.%s.%s` WHERE delivery_id = '%q'",
+		bq.projectID,
+		bq.datasetID,
+		failureEventTableID,
+		deliveryID,
+	))
+	if err != nil {
+		return false, fmt.Errorf("failed to execute FailureEventsExceedsRetryLimit: %w", err)
+	}
+
+	return res >= maxRetry, nil
+}
+
+// Write a failure event entry if there is a failure in processing the event.
+// This is used by the webhook service.
+func (bq *BigQuery) WriteFailureEvent(ctx context.Context, failureEventTableID, deliveryID, createdAt string) error {
+	inserter := bq.client.Dataset(bq.datasetID).Table(failureEventTableID).Inserter()
+	items := []*FailureEventEntry{
+		// FailureEventEntry implements the ValueSaver interface.
+		{deliveryID: deliveryID, createdAt: createdAt},
+	}
+	if err := inserter.Put(ctx, items); err != nil {
+		return fmt.Errorf("failed to execute WriteFailureEvent: %w", err)
+	}
+
+	return nil
+}
+
+// Retrieve the latest checkpoint cursor value (deliveryID) in the checkpoint table.
+// This is used by the retry service.
+func (bq *BigQuery) RetrieveCheckpointID(ctx context.Context, checkpointTableID string) (string, error) {
+	// Construct a query.
+	q := bq.client.Query(fmt.Sprintf("SELECT delivery_id FROM `%s.%s.%s` ORDER BY created DESC LIMIT 1",
+		bq.projectID,
+		bq.datasetID,
+		checkpointTableID,
+	))
+
+	// Execute the query.
+	res, err := q.Read(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to make read request to BigQuery: %w", err)
+	}
+
+	var rows []bigquery.Value
+	nextErr := res.Next(&rows)
+
+	// Table is empty, no cursor to provide
+	if errors.Is(nextErr, iterator.Done) {
+		return "", nil
+	}
+
+	if nextErr != nil {
+		return "", fmt.Errorf("failed to iterate over query response: %w", nextErr)
+	}
+
+	checkpoint, ok := rows[0].(string)
+	if !ok {
+		return "", fmt.Errorf("error converting row value (%T) to int: %v", rows[0], ok)
```

Comment:
- Nit, prefer "failed to"
- Not-nit: it probably doesn't make sense to include "ok"
- Not-nit: you failed to convert to string, not int

```suggestion
		return "", fmt.Errorf("failed to convert row value %v to string (got %T)", rows[0], rows[0])
```

Reply:
ty!